test/functional/test_framework/mininode.py

```diff
@@ -524,6 +551,33 @@ def close(self, timeout=10):
         # Safe to remove event loop.
         NetworkThread.network_event_loop = None
 
+    @classmethod
+    def listen(cls, p2p, callback, port=None, addr=None, idx=0):
+        if port is None:
+            assert idx >= 0 and idx < MAX_NODES
+            port = p2p_port(MAX_NODES-idx)
```

Comment:
I'm not sure if I'm understanding this correctly, but if you have more than `MAX_NODES / 2` nodes, could you get a port conflict?

e.g. with `MAX_NODES = 12`:

- Some code for node 7 calls `p2p_port(12 - 7)` here to get a p2p port on the Python side
- Some other code for node 5 calls `p2p_port(5)` to get a p2p port for a `bitcoind` process

Reply:
great question! however, I don't _think_ this is a problem. here's why: the `p2p_port` function in `util.py` calculates in a few different factors, one of which is the `PortSeed.n`, which is set by the `BitcoinTestFramework`. It either explicitly sets this value or defaults to the process id, which would be different for each test. so I think this would only cause a conflict if there are conflicting uses within the same test. (but I very much welcome a double check- this stuff is CONFUSING!)